### ISSUE TITLE
Bot doesn't understand its command

### ISSUE THREAD

**Issue by zabojnik** [slack-ruby/slack-ruby-bot#113]
Hi!

I run a simple pong bot but it doesn't understand its commands:
```
andy> pongbot ping
pongbot> Sorry @andy, I don't understand that command!
```

I am using this example code:
```
require 'slack-ruby-bot'

class PongBot < SlackRubyBot::Bot
  command 'ping' do |client, data, _|
    client.say(channel: data.channel, text: 'pong')
  end
end

PongBot.run
```

I registered the bot and I am running it with:
```
SLACK_API_TOKEN=CfDl12... bundle exec ruby pongbot.rb
```

Is there something basic that I am doing wrong?

Thanks!

**Comment by dblock** [comment]
Are you DMing the bot or sending a message to the bot in a channel? When you DM the bot you don't need to say `pongbot`.

**Comment by zabojnik** [comment]
I am sending a message to the bot in a channel.

**Comment by dblock** [comment]
There's one more thing, something about the bot user, does addressing the bot with `@pongbot` instead of just `pongbot` change anything?

**Comment by zabojnik** [comment]
It's the same.

This is what the log shows:

```
D, [2017-02-11T18:29:58.097650 #52325] DEBUG -- Slack::RealTime::Concurrency::Celluloid::Socket#handle_read: �~�{"type":"message","channel":"XXXX","user":"U0HTMJFPH","text":"<@U441NM1QS> ping","ts":"1486834197.000002","team":"T0HTMHL58"}
D, [2017-02-11T18:29:58.097792 #52325] DEBUG -- SlackRubyBot::Client#run_loop: WebSocket::Driver::MessageEvent, {"type":"message","channel":"XXXX","user":"U0HTMJFPH","text":"<@U441NM1QS> ping","ts":"1486834197.000002","team":"T0HTMHL58"}
D, [2017-02-11T18:29:58.098877 #52325] DEBUG -- SlackRubyBot::Client#dispatch: channel=XXXX, team=T0HTMHL58, text=<@U441NM1QS> ping, ts=1486834197.000002, type=message, user=U0HTMJFPH
D, [2017-02-11T18:29:58.102470 #52325] DEBUG -- SlackRubyBot::Client#send_json: {:type=>"message", :id=>1, :text=>"Sorry <@U0HTMJFPH>, I don't understand that command!", :channel=>"XXXX"}
D, [2017-02-11T18:29:58.107444 #52325] DEBUG -- Slack::RealTime::Concurrency::Celluloid::Socket#send_data: {"type":"message","id":1,"text":"Sorry \u003c@U0HTMJFPH\u003e, I don't understand that command!","channel":"XXXX"}
D, [2017-02-11T18:29:58.107594 #52325] DEBUG -- Slack::RealTime::Concurrency::Celluloid::Socket#write: ��00�3K�J@U�     ]�@CQ�V�ZT��VHD�        c�ABI�oE�Sp�xd�yv`�oE�U�zT�]D�F^T�ACD�]T�[QD�P_]�R^T��[Q^�V\�w�s�}�
D, [2017-02-11T18:29:58.234879 #52325] DEBUG -- Slack::RealTime::Concurrency::Celluloid::Socket#handle_read: �o{"ok":true,"reply_to":1,"ts":"1486834197.000003","text":"Sorry <@U0HTMJFPH>, I don't understand that command!"}
D, [2017-02-11T18:29:58.235000 #52325] DEBUG -- SlackRubyBot::Client#run_loop: WebSocket::Driver::MessageEvent, {"ok":true,"reply_to":1,"ts":"1486834197.000003","text":"Sorry <@U0HTMJFPH>, I don't understand that command!"}
D, [2017-02-11T18:29:58.311180 #52325] DEBUG -- Slack::RealTime::Concurrency::Celluloid::Socket#handle_read: �~�{"type":"desktop_notification","title":"COMPANY","subtitle":"pongbotdemo","msg":"1486834197.000002","content":"andy: @pongbot ping","channel":"XXXX","launchUri":"slack://channel?id=XXXX&message=1486834197000002&team=XXXX","avatarImage":"https://avatars.slack-edge.com/2016-01-15/18595913235_fa08c5e571c56807d781_192.jpg","ssbFilename":"knock_brush.mp3","imageUri":null,"is_shared":false,"event_ts":"1486834197.980109"}
D, [2017-02-11T18:29:58.311371 #52325] DEBUG -- SlackRubyBot::Client#run_loop: WebSocket::Driver::MessageEvent, {"type":"desktop_notification","title":"Dextronet","subtitle":"pongbotdemo","msg":"1486834197.000002","content":"andy: @pongbot ping","channel":"XXXX","launchUri":"slack://channel?id=XXXX&message=1486834197000002&team=XXXX","avatarImage":"https://avatars.slack-edge.com/2016-01-15/18595913235_fa08c5e571c56807d781_192.jpg","ssbFilename":"knock_brush.mp3","imageUri":null,"is_shared":false,"event_ts":"1486834197.980109"}
D, [2017-02-11T18:29:58.312733 #52325] DEBUG -- SlackRubyBot::Client#dispatch: avatarImage=https://avatars.slack-edge.com/2016-01-15/18595913235_fa08c5e571c56807d781_192.jpg, channel=XXXX, content=andy: @pongbot ping, event_ts=1486834197.980109, imageUri=, is_shared=false, launchUri=slack://channel?id=XXXX&message=1486834197000002&team=XXXX, msg=1486834197.000002, ssbFilename=knock_brush.mp3, subtitle=pongbotdemo, title=Dextronet, type=desktop_notification
```

**Comment by dblock** [comment]
I don't see anything wrong. This is unmodified code? Which version of slack-ruby-bot? You should dig this a bit deeper, try to look inside the `dispatch` method, it tries to match a command.

**Comment by fifiteen82726** [comment]
I have the same issue.

**Comment by tomchipchase** [comment]
I'm having the same issue with version `0.10.0`. Adding the command as a subclass does work, but using the DSL syntax doesn't.

I've tried with `0.9.0`, and it does work there.

**Comment by tomchipchase** [comment]
A failing spec is pretty easy to write too:

bot.rb:
```ruby
require "slack-ruby-bot"

class Bot < SlackRubyBot::Bot
  command "ping" do |client, data, _|
    client.say(text: "pong", channel: data.channel)
  end
end
```

bot_spec.rb:
```ruby
require "bot"
require "rspec-helper"

describe Bot do
  it "responds to ping" do
    expect({ message: "#{SlackBotRuby.config.user} ping")
      .to respond_with_slack_message("pong")
  end
end
```

```
Failures:

  1) Bot responds to ping
     Failure/Error: expect(message: "#{SlackRubyBot.config.user} ping").to respond_with_slack_message('pong')
     
       #<SlackRubyBot::Client:0x007fc907830ba8 @callbacks={}, @token=nil, @websocket_ping=30, @websocket_proxy=nil, @concurrency=Slack::RealTime::Concurrency::Celluloid, @start_options={}, @store_class=Slack::RealTime::Store, @logger=#<Slack::Logger:0x007fc9089b2b88 @level=2, @progname=nil, @default_formatter=#<Logger::Formatter:0x007fc9089b2b38 @datetime_format=nil>, @formatter=nil, @logdev=#<Logger::LogDevice:0x007fc9089b2ae8 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x007fc9089b2a98>>>, @web_client=#<Slack::Web::Client:0x007fc9089b28e0 @proxy=nil, @user_agent="Slack Ruby Client/0.7.9", @ca_path="", @ca_file="/ca-certificates.crt", @logger=#<Logger:0x007fc9089b2a20 @level=0, @progname=nil, @default_formatter=#<Logger::Formatter:0x007fc9089b29d0 @datetime_format=nil>, @formatter=nil, @logdev=#<Logger::LogDevice:0x007fc9089b2980 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDOUT>>, @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x007fc9089b2930>>>, @endpoint="https://slack.com/api/", @token=nil>, @aliases=nil, @send_gifs=nil> received :message with unexpected arguments
         expected: ({:channel=>"channel", :text=>"pong"})
              got: ({:text=>"Sorry <@user>, I don't understand that command!", :channel=>"channel"})
       Diff:
       @@ -1,2 +1,3 @@
       -[{:channel=>"channel", :text=>"pong"}]
       +[{:text=>"Sorry <@user>, I don't understand that command!",
       +  :channel=>"channel"}]
       
     # /usr/local/bundle/gems/slack-ruby-bot-0.10.0/lib/slack-ruby-bot/client.rb:58:in `say'
     # /usr/local/bundle/gems/slack-ruby-bot-0.10.0/lib/slack-ruby-bot/commands/unknown.rb:7:in `call'
     # /usr/local/bundle/gems/slack-ruby-bot-0.10.0/lib/slack-ruby-bot/commands/base.rb:73:in `block in invoke'
     # /usr/local/bundle/gems/slack-ruby-bot-0.10.0/lib/slack-ruby-bot/commands/base.rb:56:in `each_pair'
     # /usr/local/bundle/gems/slack-ruby-bot-0.10.0/lib/slack-ruby-bot/commands/base.rb:56:in `invoke'
     # /usr/local/bundle/gems/slack-ruby-bot-0.10.0/lib/slack-ruby-bot/hooks/message.rb:9:in `block in call'
     # /usr/local/bundle/gems/slack-ruby-bot-0.10.0/lib/slack-ruby-bot/hooks/message.rb:9:in `tap'
     # /usr/local/bundle/gems/slack-ruby-bot-0.10.0/lib/slack-ruby-bot/hooks/message.rb:9:in `call'
     # /usr/local/bundle/gems/slack-ruby-bot-0.10.0/lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_slack_message.rb:17:in `block (2 levels) in <top (required)>'
     # ./spec/bot_spec.rb:5:in `block (2 levels) in <top (required)>'

Finished in 0.09301 seconds (files took 0.58219 seconds to load)
```

**Comment by dblock** [comment]
I can confirm this is a problem. We have [a spec for this exact scenario](https://github.com/slack-ruby/slack-ruby-bot/blob/master/spec/slack-ruby-bot/commands/bot_spec.rb), but it only fails if run standalone. I'm looking into it.

**Comment by dblock** [comment]
I have a fix in https://github.com/slack-ruby/slack-ruby-bot/pull/114. Can @tomchipchase and @fifiteen82726 please confirm that it works? I'll cut a release.

**Comment by dblock** [comment]
Released 0.10.1.

**Comment by tomchipchase** [comment]
@dblock That fix works for me.

Thanks for the quick fix, and for your work on this gem!
